akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -371,6 +371,110 @@ final case class Scan[In, Out](zero: Out, f: (Out, In) ⇒ Out) extends GraphSta
 /**
  * INTERNAL API
  */
+final case class ScanAsync[In, Out](zero: Out, f: (Out, In) ⇒ Future[Out]) extends GraphStage[FlowShape[In, Out]] {
+
+  import akka.dispatch.ExecutionContexts
+
+  val in = Inlet[In]("ScanAsync.in")
+  val out = Outlet[Out]("ScanAsync.out")
+  override val shape: FlowShape[In, Out] = FlowShape[In, Out](in, out)
+
+  override val initialAttributes: Attributes = Attributes.name("scanAsync")
+
+  override val toString: String = "ScanAsync"
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with InHandler with OutHandler {
+      self ⇒
+
+      private var aggregator: Out = zero
+      private var aggregating: Future[Out] = Future.successful(aggregator)
+
+      private def ec = ExecutionContexts.sameThreadExecutionContext
+
+      private lazy val decider = inheritedAttributes.get[SupervisionStrategy].map(_.decider).getOrElse(Supervision.stoppingDecider)
+
+      private val ZeroHandler: OutHandler with InHandler = new OutHandler with InHandler {
+        override def onPush(): Unit = ()
+
+        override def onPull(): Unit = {
+          push(out, aggregator)
+          setHandlers(in, out, self)
+        }
+
+        override def onUpstreamFinish(): Unit = setHandler(out, new OutHandler {
+          override def onPull(): Unit = {
+            push(out, aggregator)
+            completeStage()
+          }
+        })
+      }
+
+      private def onRestart(t: Throwable): Unit = {
+        aggregator = zero
+      }
+
+      private def pushAndPullOrFinish(update: Out): Unit = {
+        push(out, update)
+        if (isClosed(in)) {
+          completeStage()
+        } else if (isAvailable(out) && !hasBeenPulled(in)) {
+          tryPull(in)
+        }
+      }
+
+      private val futureCB = getAsyncCallback[Try[Out]] {
+        case Success(update) if update != null ⇒
+          aggregator = update
+          pushAndPullOrFinish(update)
+        case other ⇒
+          val ex = other match {
+            case Failure(t)              ⇒ t
+            case Success(s) if s == null ⇒ ReactiveStreamsCompliance.elementMustNotBeNullException
+          }
+
+          val supervision = decider(ex)
+
+          supervision match {
+            case Supervision.Stop ⇒ failStage(ex)
+            case _ ⇒
+              if (supervision == Supervision.Restart) onRestart(ex)
+              pushAndPullOrFinish(aggregator)
```

Comment:
Doesn't this mean we push a value that may have already been pushed if resume is used?